src/mlpack/methods/ann/layer/atrous_convolution_impl.hpp

```diff
@@ -37,6 +37,41 @@ AtrousConvolution<
   // Nothing to do here.
 }
 
+template<
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename InputDataType,
+    typename OutputDataType
+>
+AtrousConvolution<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    InputDataType,
+    OutputDataType
+>::AtrousConvolution(
+  const AtrousConvolution& network) :
+    inSize(network.inSize),
+    outSize(network.outSize),
+    kernelWidth(network.kernelWidth),
+    kernelHeight(network.kernelHeight),
+    strideWidth(network.strideWidth),
+    strideHeight(network.strideHeight),
+    inputWidth(network.inputWidth),
+    inputHeight(network.inputHeight),
+    outputWidth(network.outputWidth),
+    outputHeight(network.outputHeight),
+    dilationWidth(network.dilationWidth),
+    dilationHeight(network.dilationHeight),
+    weight(network.weight),
+    bias(network.bias),
+    padding(network.padding)
+{
+  weights.set_size((outSize * inSize * kernelWidth * kernelHeight) + outSize,
+      1);
```

Comment:
Do we also need to call `Reset()` to set `weight` and `bias` pointers?

Reply:
I don't think we need to do that because when we are calling `Reset()` in a copy constructor it is leading to an extra call when using the layer with `ForwardVisitor`. Which is causing an error. I already tried it with Linear and Convolution layer

Reply:
Right, no need to call `Reset()` here, because it will be called from within the `FNN/RNN,GAN` classes.

Reply:
@zoq So, I guess, a deeper question I have is---if all of the relevant weights of a layer are set by `Reset()`, and then `Reset()` for a layer corrects any internal "alias" members, then should it be sufficient to simply copy any non-matrix members, and leave all other members alone?  (If that's true, that makes the process of making copy constructors really simple!)

Reply:
Yes, as long as you use any layer with the FNN, RNN, GAN class.

Reply:
Awesome, thanks.  @himanshupathak21061998 do you think we can try that approach then?  Basically for any copy constructor, we can ignore any matrix-type parameters, since they will be set by FFN/RNN/GAN.  I think, it should be possible to simply copy any members that are being serialized in `serialize()`.  There might be one or two exceptions, but we can figure that out as we go. :)  Let me know what you think---I think if we follow that rule, it'll vastly simplify the review and testing (and coding) process. :+1:

Personally I'm fine with the restriction that the copy constructor will then only work correctly when they are part of a network, since I don't think anyone will be using layers by themselves.

Reply:
Yes we can try that approach It will decrease the code. I have to run reset visitor for all layers in FFN/RNN/GAN.